examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,393 @@
+    image = tf.image.convert_image_dtype(image, tf.float32)
+    image = tf.image.resize(image, target_shape)
+    return image
+
+
+def generate_negative_sample(t: tf.Tensor):
+    filename = t.numpy().decode("utf-8")
+    candidate = filename
+    while candidate == filename:
+        candidate = random.choice(anchor_images).split(os.sep)[-1]
+
+    folder = random.choice([anchor_images_path, positive_images_path])
+
+    return str(folder / candidate)
+
+
+def generate_triplets(anchor_sample):
+    parts = tf.strings.split(anchor_sample, os.sep)
+    filename = parts[-1]
+
+    positive_sample = f"{str(positive_images_path)}{os.sep}" + filename
+
+    negative_sample = tf.py_function(
+        func=generate_negative_sample, inp=[filename], Tout=tf.string
+    )
+
+    return anchor_sample, positive_sample, negative_sample
+
+
+def preprocess_triplets(anchor, positive, negative):
+    return (
+        preprocess_image(anchor),
+        preprocess_image(positive),
+        preprocess_image(negative),
+    )
+
+
+"""
+Let's setup our pipeline using the list of anchor filenames as the source. The output
+of our pipeline contains a triplet with the anchor, the positive, and the negative image.
+"""
+
+dataset = tf.data.Dataset.from_tensor_slices(anchor_images)
+dataset = dataset.shuffle(buffer_size=100)
+dataset = dataset.map(generate_triplets)
+dataset = dataset.map(preprocess_triplets)
+dataset = dataset.batch(32, drop_remainder=False)
+dataset = dataset.prefetch(1)
+
+visualize(list(dataset.take(1).as_numpy_iterator())[0])
+
+"""
+## Siamese Network
+
+[Siamese Network](https://en.wikipedia.org/wiki/Siamese_neural_network) is used to solve
+many problems like detecting question duplicates, face recognition by comparing the
+similarity of the inputs by comparing their feature vectors.
+
+First we need to have a dataset that contains 3 Images, 2 are similar and 1 is different,
+they are called Anchor image, Positive Image and Negative image respectively, we need to
+tell the network that the anchor image and the positive image are similar, we also need
+to tell it that the anchor image and the negative image are NOT similar, we can do that
+by the Triplet Loss Function.
+
+Triplet Loss function:
+
+L(Anchor, Positive, Negative) = max((distance(f(Anchor), f(Positive)) -
+distance(f(Anchor), f(Negative)))**2, 0.0)
+
+Note that the weights are shared which mean that we are only using one model for
+prediction and training
+
+Also more info found here: https://sites.google.com/view/totally-looks-like-dataset
+
+Image from:
+https://towardsdatascience.com/a-friendly-introduction-to-siamese-networks-85ab17522942
+
+
+![1_0E9104t29iMBmtvq7G1G6Q.png](attachment:1_0E9104t29iMBmtvq7G1G6Q.png)
+"""
+
+"""
+First we get the paths of the datasets in siamese networks we usually have two folders
+each folder has images and every image has a corresponding similar picture in the other
+folder.
+"""
+
+
+def norm(embeddings):
```

Comment:
Consider making this a Layer rather than a function + Lambda, it's cleaner

Reply:
I added `DistanceLayer` and removed the function.